sklearn/naive_bayes.py

```diff
@@ -1528,3 +1541,600 @@ def _joint_log_likelihood(self, X):
+        return X
+
+    def _joint_log_likelihood(self, X):
+        """Calculate the meta-estimator's joint log-probability ``log P(x,y)``."""
+        estimators = self._iter(fitted=True, replace_strings=True)
+        all_jlls = Parallel(n_jobs=self.n_jobs)(
+            delayed(_jll_one)(estimator=nb_estimator, X=_safe_indexing(X, cols, axis=1))
+            for (_, nb_estimator, cols) in estimators
+        )
+        n_estimators = len(all_jlls)
+        log_prior = np.log(self.class_prior_)
+        return np.sum(all_jlls, axis=0) - (n_estimators - 1) * log_prior
+
+    def _validate_estimators(self, check_partial=False):
+        # Check if estimators have fit/partial_fit and joint log prob methods
+        # Validate estimator names via _BaseComposition._validate_names(self, names)
+        if not self.nb_estimators:
+            raise ValueError(
+                "A list of naive Bayes estimators must be provided "
+                "in the form [(name, nb_estimator, columns), ... ]."
+            )
+        names, estimators, _ = zip(*self.nb_estimators)
+        for e in estimators:
+            if (not check_partial) and (
+                not (hasattr(e, "fit") and hasattr(e, "predict_joint_log_proba"))
+            ):
+                raise TypeError(
+                    "Estimators must be naive Bayes estimators implementing "
+                    "`fit` and `predict_joint_log_proba` methods."
+                )
+            if check_partial and (
+                not (
+                    hasattr(e, "partial_fit") and hasattr(e, "predict_joint_log_proba")
+                )
+            ):
+                raise TypeError(
+                    "Estimators must be Naive Bayes estimators implementing "
+                    "`partial_fit` and `predict_joint_log_proba` methods."
+                )
+        self._validate_names(names)
+
+    def _validate_column_callables(self, X):
+        """
+        Convert callable column specifications and store into self._columns.
+
+        Empty-set columns do not enjoy any special treatment.
+        """
+        # Almost a verbatim copy of ColumnTransformer._validate_column_callables().
+        # Consider refactoring in the future.
+        # Unlike ColumnTransformer, this estimator does not need to output a
+        # dataframe or validate a the remainder, so _estimator_to_input_indices
+        # is not really needed, but retained for consistency with
+        # ColumnTransformer code.
+        all_columns = []
+        estimator_to_input_indices = {}
+        for name, _, columns in self.nb_estimators:
+            if callable(columns):
+                columns = columns(X)
+            all_columns.append(columns)
+            estimator_to_input_indices[name] = _get_column_indices(X, columns)
+        self._columns = all_columns
+        self._estimator_to_input_indices = estimator_to_input_indices
+
+    @property
+    def named_estimators_(self):
```

Comment:
I am wondering why this a property. We don't have those in stacking. We can create the estimator directly in fit once for all.

Reply:
Done. By the way, ColumnTransformer does the same thing with `named_transformers`, and I couldn't find a reason for that either.